### PR TITLE
NodeKind Refactoring

### PR DESCRIPTION
**Pull request by decanus**
As discussed, this removes the entire NodeKind class, usage of reflection and compares using instanceOf

### CODE REVIEW

File: src/Node.php [hunk 1]
```diff
@@ -312,35 +310,11 @@ protected function getChildrenKvPairs() {
     }
 
     public function jsonSerialize() {
-        $kindName = self::getNodeKindNameFromValue($this->getKind());
+        $kindName = self::getNodeKindName();
```
Comment (nikic) [inline src/Node.php]: Shouldn't this use `static::`?

File: src/Node.php [hunk 2]
```diff
@@ -15,11 +15,9 @@ class Node implements \JsonSerializable {
...
     public $parent;
 
-    public function getKind() {
-        // TODO remove all references to getKind
-        $constants = (new \ReflectionClass("Microsoft\\PhpParser\\NodeKind"))->getConstants();
-        $nameParts = explode("\\",get_class($this));
-        return $constants[end($nameParts)];
+    public function getNodeKindName() : string {
+        $parts = explode('\\', static::class);
+        return $parts[count($parts) - 1];
```
Comment (nikic) [inline src/Node.php]: An alternative implementation would be to use strrpos(), which avoid creating a temporary array.

File: src/Node.php (hunk 1 above)
Comment (mousetraps) [inline src/Node.php]: Good catch - it doesn't make much of a functional difference at the moment, but might help if we ever decide to override `getNodeKindName`

**Comment by mousetraps** [comment]
FYI pushed some minor changes to the branch, as per @nikic's comments.

**Comment by decanus** [comment]
@mousetraps Thanks for making the changes

**Pull request merged**